File: leet_BFS_0222_count_node.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function countNodes(root: TreeNode | null): number {
    if ( !root ) {
        return 0
    }
    let D: TreeNode[] = [ root ]
    let res: number = 0
    while (D.length) {
        let i = -1
        while ( ++i < D.length ) {
            let node: TreeNode = D.shift()
            if (node.left) {
                D.push(node.left)
            }
            if (node.right) {
                D.push(node.right)
            }
            ++res
        }
    }
    return res
};

File: leet_BFS_0222_count_node.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}

	D := []*TreeNode{ root }
	res := 0

	for len(D) > 0 {

		size := len(D)
		for i := 0; i < size; i++ {
			node := D[0]
            D = D[1:]
			if node.Left != nil {
				D = append(D, node.Left)
			}
			if node.Right != nil {
				D = append(D, node.Right)
			}
			res++
		}
	}

	return res
}
